source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -569,9 +627,9 @@ void Filter::onUpstreamEvent(Network::ConnectionEvent event) {
         read_callbacks_->upstreamHost()->outlierDetector().putResult(
             Upstream::Outlier::Result::LocalOriginConnectFailed);
       }
-
       if (!downstream_closed_) {
-        initializeUpstreamConnection();
+        route_ = pickRoute();
```

Comment:
Please add a test case that validates that this re-pick happens

Reply:
ack

Reply:
add PickClusterOnUpstreamFailure and OnDemandCallbackStickToTheSelectedCluster 
to verify the re-pick at both normal and on demand